Add canvas zoom

barklund opened the pull request:
## Context

This adds the ability to specify the page zoom in the editor allowing users to zoom in to a page to adjust details regardless of their browser window size.

## Summary

The possible zoom levels are two adaptive ones and two fixed-size ones:

- **Fit**: Adaptive - makes sure the entire page is visible and there will never be scrollbars.
- **Fill**: Adaptive - makes sure that the entire workspace is used at the smallest possible zoom level.
- **100%**: Fixed-size.
- **200%**: Fixed-size.

If the page at any given zoom level requires scrollbars, these will be displayed and a tiny preview of the currently visible viewport will be visible in the carousel.

If there's room for page navigation on the sides of the page, they'll be displayed. Otherwise not.

## Relevant Technical Choices

- Added new functionality to the top-level layout provider
- Removed some functionality from the canvas provider
- Had some issues with the layer system, but think I managed to get it working without breaking other stuff

## To-do

- [x] Add dropdown
- [x] Extend layout provider with necessary functionality
- [x] Make canvas allow overflowing page
- [x] Make canvas layers follow top frame layer
- [x] Add zoom viewport preview in the carousel
- [x] Fix various overflow bugs
- [x] Add scroll bar gutter
- [x] Fix existing tests
- [x] Update to new designs
- [x] Add experimental flag and use comparable "old" logic if flag is false
- [x] Fix remaining failing tests
- [x] File new bug for lasso select
- [x] File new bug for adding tests

## User-facing changes

A couple of video demos:

_To be inserted_


## Testing Instructions

### QA

This PR can be tested by following these steps:

1. _To be defined_

### UAT

<!--
Sometimes the testing instructions for UAT can differ from the ones for QA.
-->

<!-- ignore-task-list-start -->
- [x] UAT should use the same steps as above.
<!-- ignore-task-list-end -->

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #4022

Commit:
Moved carousel to new component folder along with relevant files

Commit:
Refactored carousel into many components

Commit:
Fix carousel menu icons

Commit:
Fixed grid view page preview reference

Commit:
Almost fixed page borders - old style clashed with round corners

Commit:
Fixed semantics of elements

Commit:
Fix all the ref listeners - must re-run when ref is actually populated

Commit:
Fixed rearrange bug

Commit:
Merge branch 'main' into add/5692-new-carousel

# Conflicts:
#	assets/src/dashboard/utils/index.js
#	assets/src/design-system/components/keyboard/gridview/test/getIndex.js
#	assets/src/edit-story/components/canvas/carousel/index.js
#	assets/src/edit-story/components/canvas/gridview/test/useGridViewKeys.js
#	assets/src/edit-story/components/carousel/gridview/test/useGridViewKeys.js
#	assets/src/edit-story/components/carousel/gridview/useGridViewKeys.js
#	assets/src/edit-story/components/carousel/useCarouselKeys.js
#	assets/src/edit-story/components/keyboard/index.js

Commit:
Removed unused file

Commit:
Moved constants to separate file

Commit:
Renamed prop

Commit:
Removed unused prop

Commit:
Simplified function

Commit:
Implemented much cleaner scroll fix - and looks bloody nice!

Commit:
Re-add label for tests

Commit:
Fixed esc2blur for direct node references

Commit:
Fixed re-setting focus to new active page, when a page is deleted

Commit:
Reverted stuff that removed arrow-navigation focus

Commit:
Hidden arrows when no scroll is available

Commit:
Fixed some karma test

I have no idea why this started failing, but it did so consistently. The handle was not clickable in the top left corner at (1,1), so should never have worked?

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Added new (dumb) zoom level selector

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Initial fully working version with a couple bugs still

Commit:
Fixed some bugs referring to layout provider

Commit:
Added missing files - and fixed some more layout issues

Commit:
Reverted overflow - cannot change this

Commit:
Renamed to offset and made it a pure number

Commit:
Added a zoom preview

This was actually not in the spec, but it felt necessary when working with the feature. And it is pretty awesome!

Commit:
Fixed weird overflow error

Commit:
Fixed another global overflow bug

This only happening when snapping to a large item while dragging directly from library. But it shifted everything around. I hope this is safe. @swissspidy?

barklund commented on assets/src/edit-story/style.css:
```diff
@@ -39,6 +39,7 @@ body.js.edit-story #web-stories-editor {
   background-color: #1b1d1c; /* theme.colors.bg.workspace */
   position: relative;
   height: calc(100vh - 32px); /* ADMIN_TOOLBAR_HEIGHT = 32 */
+  overflow: hidden;
```
This one felt a bit dirty, but there was no other way around it. If you zoomed in to 300%, added a large image and then tried to drag in another image from library and tried snapping to the large image on screen, the snap guides would exceed the workspace and cause the entire browser window to have scrollbars. I couldn't find a good location in the editor to set this overflow, so this seemed to be the best location. It'll also fix any other weird overflow issues we might get in the future.

But I'm curious if anyone can see a problem with this?

miina replied:
It shouldn't cause a problem at this point, the only thing it could do is hide some potential bugs with overflow that would require fixing in another way. There has been at least one overflow case where unused Moveable elements were left outside of the editor view and this caused overflow. Hiding overflow would have "fixed" it but it actually needed cleaning up the unused Moveables instead.

swissspidy replied:
I tried to break it, but couldn't. So far so good I suppose.

miina commented on assets/src/edit-story/constants/index.js:
```diff
@@ -28,6 +28,14 @@ export const PAGE_NAV_PADDING = 60;
 export const PAGE_NAV_WIDTH = PAGE_NAV_PADDING + PAGE_NAV_BUTTON_SIZE;
 
+export const ZOOM_SETTING = {
+  SINGLE: 'SINGLE',
```
There's also the pinching functionality in Figma ("Pinch with two fingers to zoom in and out") which seems to set the numbers to custom %. If that's also coming, should the `SINGLE` perhaps represent `100%`, `DOUBLE` `200%` etc. so the values would be used dynamically to set the zoom to use the same system as the custom % will use? Thoughts?

barklund replied:
I have some ideas about how to refactor when getting to pinch, but it'll be moved to a future ticket and I'll deal with it then. I'm thinking a value of `FIXED` and a separate value for the current zoom percentage. Because if you pinch while at `FIT` or `FILL`, you still need to know the current zoom percentage to increase or decrease it (and change setting to `FIXED`), so it's needed always.

barklund replied:
Formulated thoughts in #6179.

miina commented on assets/src/edit-story/app/layout/useZoomSetting.js:
```diff
@@ -0,0 +1,114 @@
+        }
+        break;
+      }
+      case ZOOM_SETTING.FIT: {
```
Is it intended that the FIT option has no minimum value and can go infinitely small?

barklund replied:
Yes 🤏

miina commented on assets/src/edit-story/components/canvas/layout.js:
```diff
@@ -195,7 +243,11 @@ const PageArea = forwardRef(function PageArea(
+      <PageAreaWithOverflow
+        showOverflow={showOverflow}
+        background={background}
+        isControlled={isControlled}
```
What's the `isControlled` for? Is that for marking the zoomable area?

barklund replied:
It's to say that the two passive layers ([display](https://github.com/google/web-stories-wp/pull/6112/files/92f317c9b78e25dde830bf1e73febdaa753cbbdb#diff-54b708c892748c7c0910902d490dfdb7bb641c483b3a3f223a636deb7d89294aR165) and [edit](https://github.com/google/web-stories-wp/pull/6112/files/92f317c9b78e25dde830bf1e73febdaa753cbbdb#diff-92a351d00597f253df518e2916506e81677678e6f65ef613f6e49c133cb0111dR101)) are "controlled" by the scrolling manually happening in the frame layer: event [here](https://github.com/google/web-stories-wp/pull/6112/files/92f317c9b78e25dde830bf1e73febdaa753cbbdb#diff-7f013d952826974508394ad2b82986572fbaa0ad58b324d36d6cd48a0bf1db46R111-R119), CSS var [here](https://github.com/google/web-stories-wp/pull/6112/files/92f317c9b78e25dde830bf1e73febdaa753cbbdb#diff-f634fcb15355d7101d454a9ad3b2aa06ceb1ce6bd5ebd0f4e8dcf55c9ded3441R223-R224), CSS var usage [here](https://github.com/google/web-stories-wp/pull/6112/files/92f317c9b78e25dde830bf1e73febdaa753cbbdb#diff-92a351d00597f253df518e2916506e81677678e6f65ef613f6e49c133cb0111dR101).

miina replied:
Ok, I wonder if there's a better name for this to hint that it's related to scrolling. `isControlled` could mean different things without context. Thoughts?

barklund replied:
This has to be re-worked when fixing #6180, so leaving it as-is for now.

miina commented on assets/src/edit-story/components/canvas/framesLayer.js:
```diff
@@ -98,15 +101,28 @@ function FramesLayer() {
     <Layer
       ref={ref}
       data-testid="FramesLayer"
-      pointerEvents="none"
```
Hm, I wonder if it was `none` before for a specific purpose.

barklund replied:
Ditto, but couldn't find any problem with removing it. And it's needed to allow scroll events. Drag-to-select and everything else still seems to work. It's hopefully something that was necessary once, but not anymore.

miina replied:
Has this PR merged clicking through the empty area of a shape (#2038)? Could you merge `main` to ensure we're testing up to date features?

barklund replied:
I just merged `main` in. Let me know if you experience regressions.

miina replied:
Hm, I’m not able to access image edit mode currently, working ok for you?

miina replied:
![edit](https://user-images.githubusercontent.com/3294597/106620815-3aac9c80-6572-11eb-99b7-35029830cd6c.gif)

miina replied:
^ Ignore this, seems to happen on `main`, too!

barklund replied:
So, this original line of code did have an impact on the video control element, which [needed a manual `pointer-events: none`](https://github.com/google/web-stories-wp/pull/6112/commits/2e80ecebca8ebb7f4da657920cee09e2903c2856), but other than that, everything else seems to work.

miina reviewed (commented):
Nice! Thanks for the helpful explanatory comments in the code, too. Left some questions in the code and some notes from testing as well in a separate comment but generally looks great!

Commit:
Fixed overflow bug, but might have caused regression?

Commit:
Moved zoom preview out from mask

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Set generic scrollbar corner color globally

It is only used when scrolling in two directions, which we rarely do, but needed now for the canvas.

Commit:
Cleaned it up a bit

Commit:
Fixed edit layer scrolling

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Finally fixed the weird scrollbar gutter issues

Commit:
Fixed unit tests

Commit:
Fixed bug around missing inherited css property

miina commented on assets/src/edit-story/app/canvas/canvasProvider.js:
```diff
@@ -43,9 +43,9 @@ function CanvasProvider({ children }) {
-  const { pageSize, setPageSize } = useLayout(({ state, actions }) => ({
-    pageSize: state.canvasPageSize,
-    setPageSize: actions.setCanvasPageSize,
+  const pageSize = useLayout(({ state: { pageWidth, pageHeight } }) => ({
```
Should `useLayout` be used everywhere for getting the `pageSize` perhaps and `pageSize` removed from returning from `CanvasProvider` at all since it got replaced in most of the places? `useCanvas` is currently still used in one place only (in `pagemenu/index.js`). Thoughts?

barklund replied:
Hm, that's a good point. Didn't even check how many instances there were of this. Fixed in 6bd6863a3 with follow-up in f48462ee9 and 68bca1504.

miina commented on assets/src/edit-story/app/layout/useZoomSetting.js:
```diff
@@ -0,0 +1,116 @@
+        break;
+    }
+    // Floor page width to nearest multiple of 12
+    const pageWidth = 12 * Math.floor(maxPageWidth / 12);
```
12 to a constant?

barklund replied:
Done in 28ec182bb. Might be [soft coding](https://thedailywtf.com/articles/Soft_Coding) though, but works for me.

miina commented on assets/src/edit-story/components/canvas/test/_utils.js:
```diff
@@ -111,6 +118,13 @@ export function TestDisplayElement({
+  const layoutContext = {
+    state: {
+      pageWidth: 412,
+      pageHeight: 618,
```
Maybe move to a constant?

barklund replied:
Already are, but now they're imported in 270b05da9.

miina reviewed (commented):
LGTM. Will approve with the tests!

One more note from testing, too:
The frame line is on top of the scrollbar, since the scrollbar is almost invisible (at least to my eye) then the frame being on top of it leaves the impression as if the page has ended but actually it hasn't:
![scrolling](https://user-images.githubusercontent.com/3294597/106872492-a8240e80-66d3-11eb-8e02-f79826c17c1c.gif)

Thoughts? Can probably be addressed later.

Commit:
Merge branch 'main' into add/4022-zoom

swissspidy commented on assets/src/edit-story/components/carousel/zoomSelector/zoomSelector.js:
```diff
@@ -0,0 +1,73 @@
+  { name: '200%', value: ZOOM_SETTING.DOUBLE },
+  { name: '300%', value: ZOOM_SETTING.TRIPLE },
+  { name: 'Fill', value: ZOOM_SETTING.FILL },
+  { name: 'Fit', value: ZOOM_SETTING.FIT },
```
The names should be translatable. Ideally with context.

```suggestion
  { name: _x('100%', 'zoom level', 'web-stories'), value: ZOOM_SETTING.SINGLE },
  { name: _x('200%', 'zoom level', 'web-stories'), value: ZOOM_SETTING.DOUBLE },
  { name: _x('300%', 'zoom level', 'web-stories'), value: ZOOM_SETTING.TRIPLE },
  { name: _x('Fill', 'zoom level', 'web-stories'), value: ZOOM_SETTING.FILL },
  { name: _x('Fit', 'zoom level', 'web-stories'), value: ZOOM_SETTING.FIT },
```

Commit:
Fixed i18n

Co-authored-by: Pascal Birchler <pascalb@google.com>

Commit:
Merge branch 'add/4022-zoom' of github.com:google/web-stories-wp into add/4022-zoom

Commit:
Fixed imports

Commit:
Fixed constants in canvas test utils

Commit:
Removed unused constant and added new

Commit:
Changed refs to page size to refer to layout provider

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Removed page size completely from canvas provider

Commit:
Updated references to `useLayout`

Commit:
Removed irrelevant code - that scale has always just been 1

Commit:
Fixed bug in snapping

Commit:
Merge branch 'main' into add/4022-zoom

Commit:
Removed zoom preview

swissspidy committed:
Merge branch 'main' into add/4022-zoom

swissspidy commented on assets/src/edit-story/components/carousel/zoomSelector/zoomSelector.js:
```diff
@@ -0,0 +1,73 @@
+ * WordPress dependencies
+ */
+import { __, _x } from '@wordpress/i18n';
+
```
```suggestion
import { __, _x } from '@web-stories-wp/i18n';
```

swissspidy committed:
Lint fix

Commit:
Adjust zoom (#6236)

* Improved spacing around page menu

* Reworked menu spacing and add fit padding

* Added padding to page when double-scrolling

* Remove triple option

* Added extra padding for higher zoom levels

Commit:
Initial merge, still a few bugs

Commit:
Fixed page nav

Commit:
Fixed scrollbar height

Commit:
Complete overhaul of zoom viewport logic

Commit:
More rework of overflow logic

Commit:
Maybe finally fixed overflow stuff?

Commit:
Fixed frame updates on zoom change

Commit:
Revert autofix

miina commented on assets/src/edit-story/components/carousel/zoomSelector/zoomSelector.js:
```diff
@@ -0,0 +1,67 @@
+  );
+
+  return (
+    <DropDown
```
Is this purposefully using the old DropDown currently?

barklund replied:
Oh, forgot about that. Will create follow-up ticket for that.

barklund replied:
Filed in #6912.

miina commented on assets/src/edit-story/components/canvas/singleSelectionMoveable/index.js:
```diff
@@ -71,6 +78,25 @@ function SingleSelectionMoveable({
   }, [pushEvent]);
 
+  // If scroll ever updates, update rect now
+  useEffect(() => {
```
Maybe it would be good to create a hook for it later. This will also be needed for the Moveable used in the edit mode later when Zoom will work for edit mode, too.

miina commented on assets/src/edit-story/components/canvas/layout.js:
```diff
@@ -185,71 +243,148 @@ const CarouselArea = styled(Area).attrs({
-          {children}
-        </PageAreaSafeZone>
-      </PageAreaWithOverflow>
+      <PageClip
```
Looks like something is cutting off the Media edit mode now -- it should have overflow allowed and the Zoom slider visible, currently it's hidden. Perhaps the `PageClip` is doing that?
![overflow](https://user-images.githubusercontent.com/3294597/112302677-a0c4be80-8c9b-11eb-9fc3-a52c0e24b616.gif)

barklund replied:
Fixed it - was tricky!

miina commented on assets/src/edit-story/components/canvas/layout.js:
```diff
@@ -115,45 +115,103 @@ const PageAreaFullbleedContainer = styled(Area).attrs({
+        right: -4px;
+        bottom: -4px;
+        border: ${theme.colors.border.selection} 1px solid;
+        border-radius: ${theme.borders.radius.medium};
```
👍

miina commented on assets/src/edit-story/app/layout/useOffset.js:
```diff
@@ -0,0 +1,37 @@
+ */
+import { useState, useEffect } from 'react';
+
+function useOffset(zoomSetting) {
```
Maybe rename this to `useScrollOffset` for clarity?

barklund replied:
Done in 96363564a.

miina commented on assets/src/design-system/theme/helpers/scrollbar.js:
```diff
@@ -32,6 +32,7 @@ export const scrollbarCSS = () => css`
 
   ::-webkit-scrollbar {
     width: 11px;
```
Should these use the scrollbar width constants?

barklund replied:
Then it should be moved to the theme, but yeah, makes sense.

barklund replied:
Moved in 60b9275e2.

miina reviewed (commented):
Great effort here, works well -- good idea to have Zoom behind a feature flag and implementing the rest separately! 🎉 

Found two issues that are blockers unless these would not happen behind feature flag:
1. The media edit mode overflow is cut off from the Page (gif in the comments)
2. Lasso selection doesn't seem to work correctly -- it's making the selection based on non-zoomed position, I think (Text gets selected instead of the Shape):
![lasso](https://user-images.githubusercontent.com/3294597/112303823-09606b00-8c9d-11eb-95eb-f59b78ef03f0.gif)

Commit:
Positione edit layer without any overflow ever

Commit:
Renamed hook

miina commented on assets/src/edit-story/components/canvas/layout.js:
```diff
@@ -120,13 +120,13 @@ const PageAreaFullbleedContainer = styled(Area).attrs({
     isControlled &&
     css`
-      overflow: hidden;
+      overflow: ${({ overflow }) => overflow ?? 'hidden'};
```
Hm, couldn't this also come from the `showOverflow`? Or is there a possibility where `showOverflow === false` but `overflow === 'visible'`? Thoughts?

barklund replied:
Don't ask! 😂 

I've spent too much time on this already 🙈 . It could probably be simplified - or maybe even split into three completely different layers now just using the same grid, as the three layers are actually pretty different by now 🤷 . But I'll leave that as an exercise for the reader...

miina replied:
Haha! Mb this can be followed up in another PR 😆 Seems like `showOverflow` means `overflow: visible` and `!showOverflow` means `visibility: hidden` (at least that was the original idea for `showOverflow`) so unless I'm missing something then `overflow` might be duplicating the point of `showOverflow`.

Commit:
Added feature flag and re-implemented old zoom logic

Commit:
Fixed and generalized hook to update based on scroll

Commit:
Merge branch 'main' into add/4022-zoom

# Conflicts:
#	includes/Experiments.php

Commit:
Fixed various tests dragging in bg images

Also disabled tests for lasso selection, as it is slightly broken currently, to be fixed in a future PR.

Commit:
Moved scrollbar constant to where it belongs

barklund commented on assets/src/edit-story/elements/media/scalePanel.js:
```diff
@@ -45,6 +45,7 @@ const Container = styled.div`
   border-radius: 8px;
   padding: 3px ${HORIZONTAL_PADDING}px;
   margin-top: 8px;
+  box-shadow: 0 0 3px white;
```
This should use a design-system color but awaiting confirmation that this is actually okay to add.
